aesm-client/src/imp/aesm_protobuf/mod.rs

```diff
@@ -107,4 +107,146 @@ impl AesmClient {
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        let res = self.transact(req)?;
+        let num_key_ids : u32 = res.get_att_key_id_num();
+
+        let mut req = Request_GetSupportedAttKeyIDsRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        // sgx defined protobuf returns an opaque byte array with size we give it. Byte array MUST have expected size of structure size*num.
+        let expected_buffer_size : u32 = num_key_ids * SGX_KEY_ID_SIZE;
+        req.set_buf_size(expected_buffer_size);
+
+        // fetch requested data
+        let mut res = self.transact(req)?;
+
+        // Interpret the byte array as defined in C header at linux-sgx/common/inc/sgx_quote.h, this is converted to sgx_quote.rs via bindgen.
+        let key_ids_blob = res.take_att_key_ids();
+
+        if key_ids_blob.len() as u32 != expected_buffer_size {
+            return Err(Error::aesm_bad_response("wrong buffer size returned by aesm service"));
+        }
+
+        let mut key_ids : Vec<Vec<u8>> = Vec::new();
+        
+        for i in key_ids_blob.chunks_exact(SGX_KEY_ID_SIZE as usize) {
+            key_ids.push(i.to_vec());
+        }
+        
+        Ok(key_ids)
+    }
+    
+    ///
+    /// Returns information needed by an intel SGX application to get a quote of one of its enclaves.
+    /// Similar functionality to sgx_init_quote_ex in page 165 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// att_key_id - a byte array with hidden type 'sgx_att_key_id_ext_t' returned by AesmClient::select_att_key_id or AesmKeyIds::select_algorithm_id or similar.
+    ///
+    /// returns information to use for get_quote_ex
+    ///
+    /// note - intentionally skipping p_pub_key_id and buf_size as they are not useful at the moment.
+    ///        a) buf_size must be exactly sizeof(sgx_sha256_hash_t) otherwise fail.
+    ///        b) p_pub_key_id set to false causes response to be empty except 'pub_key_id_size'. But we must already know that and set it via buf_size - otherwise see point a.
+    ///
+    pub fn init_quote_ex(&self, att_key_id: Vec<u8>) -> Result<QuoteInfoEx> {
+        const SGX_SHA256_SIZE : u64 = 32;
+        let mut req = Request_InitQuoteExRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        req.set_att_key_id(att_key_id);
+        req.set_b_pub_key_id(true);
+        req.set_buf_size(SGX_SHA256_SIZE);
```

Comment:
Shouldn't you be calling this API twice, the first time to get this size?

Reply:
That also works, updated to do that.